### PR TITLE
Add RDT Memory Bandwidth Monitoring (MBM)

### PR DESCRIPTION
**Pull request by Creatone**

### CODE REVIEW

File: libcontainer/intelrdt/intelrdt.go [hunk 1]
```diff
@@ -217,20 +221,31 @@ func init() {
...
 }
 
 // Return the mount point path of Intel RDT "resource control" filesysem
 func findIntelRdtMountpointDir() (string, error) {
 	f, err := os.Open("/proc/self/mountinfo")
+
 	if err != nil {
 		return "", err
 	}
-	defer f.Close()
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/intelrdt.go:233]: Why did you remove closing of the file? I think that it's not a good idea.

File: libcontainer/intelrdt/intelrdt.go [hunk 2]
```diff
@@ -192,7 +196,7 @@ type intelRdtData struct {
 func init() {
 	// 1. Check if hardware and kernel support Intel RDT sub-features
 	// "cat_l3" flag for CAT and "mba" flag for MBA
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/intelrdt.go:198]: you have changed `parseCpuInfoFile ` function so please update this comment

File: libcontainer/intelrdt/mbm.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	llcOccupancy  bool
+}
+
+// Check if Intel RDT/MBM is enabled
+func IsMbmEnabled() bool {
+	return isMbmEnabled
+}
+
+func getMonFeatures(intelRdtRoot string) (monFeatures, error) {
+	file, err := os.Open(filepath.Join(intelRdtRoot, "info", "L3_MON", "mon_features"))
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/mbm.go:31]: this file is never closed

File: libcontainer/intelrdt/mbm.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	return parseMonFeatures(file)
+}
+
+func parseMonFeatures(file io.Reader) (monFeatures, error) {
+	s := bufio.NewScanner(file)
+
+	monFeatures := monFeatures{}
+
+	for s.Scan() {
+		if err := s.Err(); err != nil {
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/mbm.go:44]: please double check if it is a good idea to check scan errors in the loop, to my knowledge if any error appears during scanning then the loop is broken, I used to see following form of scan loop
```
for scanner.Scan() {
    fmt.Println(scanner.Text())
}
if scanner.Err() != nil {
    fmt.Println(scanner.Err())
}
```

File: libcontainer/intelrdt/mbm.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+		switch feature := s.Text(); feature {
+
+		case "mbm_total_bytes":
+			monFeatures.mbmTotalBytes = true
+		case "mbm_local_bytes":
+			monFeatures.mbmLocalBytes = true
+		case "llc_occupancy":
+			monFeatures.llcOccupancy = true
+		}
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/mbm.go:56]: I would like to add here a default case to catch unsupported/unexpected feature. What do you think?

File: types/events.go [hunk 6]
```diff
@@ -1,5 +1,9 @@
 package types
 
+import (
```
Comment (katarzyna-z) [inline types/events.go:3]: if there is only one import you can write simply 
```
import  "github.com/opencontainers/runc/libcontainer/intelrdt"
```

File: events.go [hunk 7]
```diff
@@ -218,15 +218,15 @@ func convertMemBwInfo(i *intelrdt.MemBwInfo) *types.MemBwInfo {
 }
 
 func convertMbmNumaNodeStats(i *[]intelrdt.MbmNumaNodeStats) *[]types.MbmNumaNodeStats {
-	mbmNumaNodeStats := make([]types.MbmNumaNodeStats, 0)
+	mbmNumaNodeStats := make([]types.MbmNumaNodeStats, len(*i))
 
-	for _, stats := range *i {
+	for place, stats := range *i {
```
Comment (iwankgb) [inline events.go]: `index` might be better name (this is what traditional `i` stands for).

File: libcontainer/intelrdt/intelrdt.go [hunk 8]
```diff
@@ -217,16 +220,28 @@ func init() {
...
+		if err != nil {
+			return
+		}
+	}
 }
 
 // Return the mount point path of Intel RDT "resource control" filesysem
 func findIntelRdtMountpointDir() (string, error) {
 	f, err := os.Open("/proc/self/mountinfo")
+
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/intelrdt.go:244]: It is not needed to add this empty line here.

File: libcontainer/intelrdt/mbm.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
+// +build linux
+
+package intelrdt
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/sirupsen/logrus"
```
Comment (katarzyna-z) [inline libcontainer/intelrdt/mbm.go:8]: Usually standard libraries are split from non-standard packages on the list of imports, proper form of imports have following form:
```
import (
        "bufio"
	"fmt"
	"io"
	"os"
	"path/filepath"

        "github.com/sirupsen/logrus"
)
```
goimport can also help with formatting list of imports, see the section "Formatting Imports" in [this ](https://www.digitalocean.com/community/tutorials/importing-packages-in-go) document.

File: libcontainer/intelrdt/mbm_test.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+			}
+
+			if numaStats.LLCOccupancy != mocksFilesToCreate["llc_occupancy"] {
+				t.Fatal("Wrong value of llc_occupancy")
+			}
+		}
+
+	})
+
+	err = os.RemoveAll(mockedMBM)
```
Comment (iwankgb) [inline libcontainer/intelrdt/mbm_test.go:132]: This call should be deferred immediately after creating the file structure.

File: libcontainer/intelrdt/mbm_test.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+		stats, err := getMBMStats(mockedMBM)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, numaStats := range *stats {
+
+			if numaStats.MBMTotalBytes != mocksFilesToCreate["mbm_total_bytes"] {
+				t.Fatal("Wrong value of mbm_total_bytes")
```
Comment (iwankgb) [inline libcontainer/intelrdt/mbm_test.go:118]: Improve the messages by stating what was expected value and what was actual value.

File: libcontainer/intelrdt/mbm_test.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+		enabledMonFeatures.mbmTotalBytes = true
+		enabledMonFeatures.mbmLocalBytes = true
+		enabledMonFeatures.llcOccupancy = true
+
+		stats, err := getMBMStats(mockedMBM)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, numaStats := range *stats {
```
Comment (iwankgb) [inline libcontainer/intelrdt/mbm_test.go:115]: Do not iterate, assert slice length and use an index to read values, please.
